File: typescript/src/n0141_linked_list_cycle.test.ts
import { expect } from "jsr:@std/expect";
import { hasCycle } from "./n0141_linked_list_cycle.ts";
import { List } from "./utils/linked_list.ts";

Deno.test("hasCycle example 1", () => {
  const list = new List([3, 2, 0, -4]);

  const nodes = Array.from(list.nodes());
  nodes[nodes.length - 1]!.next = nodes[1]!;

  expect(hasCycle(list.head)).toBe(true);
});

Deno.test("hasCycle example 2", () => {
  const list = new List([1, 2]);

  const nodes = Array.from(list.nodes());
  nodes[nodes.length - 1]!.next = nodes[0]!;

  expect(hasCycle(list.head)).toBe(true);
});

Deno.test("hasCycle example 3", () => {
  const list = new List([1]);
  expect(hasCycle(list.head)).toBe(false);
});

File: rust/src/n0141_linked_list_cycle.test.rs
use crate::n0141_linked_list_cycle::*;
use crate::utils::linked_list::SharedList;

#[test]
fn test_example_one() {
    let list = SharedList::new([3, 2, 0, -4]);
    let nodes = list.nodes().collect::<Vec<_>>();
    nodes[nodes.len() - 1].borrow_mut().next = Some(nodes[1].clone());
    assert!(has_cycle(list.head));
}

#[test]
fn test_example_two() {
    let list = SharedList::new([1, 2]);
    let nodes = list.nodes().collect::<Vec<_>>();
    nodes[nodes.len() - 1].borrow_mut().next = Some(nodes[0].clone());
    assert!(has_cycle(list.head));
}

#[test]
fn test_example_three() {
    let list = SharedList::new([1]);
    assert!(!has_cycle(list.head));
}
